### PR TITLE
Fallout

### PR DESCRIPTION
**Pull request by tommbendall**
This introduces a physics routine that performs advection of a rain field. The rain is designed to fall out of the bottom of the domain.

### CODE REVIEW

File: gusto/diagnostics.py [hunk 1]
```diff
@@ -401,12 +401,47 @@ def name(self):
...
+        self.total_water.assign(self.total_water + self.flux)
+        return self.field.interpolate(self.total_water)
+
+
 class Vorticity(DiagnosticField):
     """Base diagnostic field class for vorticity."""
 
     def setup(self, state, vorticity_type=None):
         """Solver for vorticity.
-
```
Comment (dorugeber) [inline gusto/diagnostics.py]: ???

File: gusto/diagnostics.py [hunk 2]
```diff
@@ -461,7 +496,6 @@ def setup(self, state):
         a weighted mass system to generate the
         potential vorticity from known velocity and
         depth fields.
-
```
Comment (dorugeber) [inline gusto/diagnostics.py]: ???

File: gusto/diagnostics.py [hunk 3]
```diff
@@ -473,7 +507,6 @@ class AbsoluteVorticity(Vorticity):
 
     def setup(self, state):
         """Solver for absolute vorticity.
-
```
Comment (dorugeber) [inline gusto/diagnostics.py]: ???

File: gusto/diagnostics.py [hunk 4]
```diff
@@ -485,7 +518,6 @@ class RelativeVorticity(Vorticity):
 
     def setup(self, state):
         """Solver for relative vorticity.
-
```
Comment (dorugeber) [inline gusto/diagnostics.py]: ???

File: gusto/physics.py [hunk 5]
```diff
@@ -1,15 +1,16 @@
...
 
 
-__all__ = ["Condensation"]
+__all__ = ["Condensation", "Fallout"]
 
 
 class Physics(object, metaclass=ABCMeta):
     """
     Base class for physics processes for Gusto.
-
```
Comment (dorugeber) [inline gusto/physics.py]: ???

File: gusto/physics.py [hunk 6]
```diff
@@ -31,7 +32,6 @@ class Condensation(Physics):
     into liquid water and evaporation of liquid
     water into water vapour, with the associated
     latent heat changes.
-
```
Comment (dorugeber) [inline gusto/physics.py]: ???

File: gusto/diagnostics.py (hunk 1 above)
Comment (tommbendall) [reply gusto/diagnostics.py]: Ah sorry, it wasn't deliberate. I had copied and pasted this code over from a different branch that had other features and somehow copying into the emacs window has deleted these lines!

**Comment by dorugeber** [comment]
FWIW, the reason I took issue to this is that it suggests you aren't running `git diff` before committing changes?  If so, you're not really "in control" of what you're actually committing.  @miklos1 would tell you to run `git status` and `git diff` like you have some OCD about it.

**Comment by miklos1** [comment]
Yes, some OCDs should be called disciple, and those ones are a good thing.

Yes, if you don't check the diff before committing, then you basically don't know what you are doing.

File: gusto/transport_equation.py [hunk 7]
```diff
@@ -178,6 +186,10 @@ def advection_term(self, q):
                             dot(self.ubar('+'), self.n('+'))*q('+'))
                       + inner(self.test('-'),
                               dot(self.ubar('-'), self.n('-'))*q('-')))*self.dS
+
+        if self.outflow and self.ibp is not None:
```
Comment (jshipton) [inline gusto/transport_equation.py]: what happens if `ibp` is not None and `self.inflow` is `True`? Should this warn the user that they have set incompatible options?

Comment (tommbendall) [reply gusto/transport_equation.py]: Ah have I written this a bit confusingly? This means that if `self.outflow is True` and if `ibp is not None` then we need to add the surface integrals on the edge of the domain.

And I think that if you set `ibp=None` then you get both inflow and outflow, so I think this is the only case we need to consider?

File: gusto/diagnostics.py [hunk 8]
```diff
@@ -401,6 +401,42 @@ def name(self):
...
+            self.flux = Function(space)
+            self.total_water = Function(space)
+
+            if space.extruded:
+                self.ds = ds_b + ds_t + ds_v
+            else:
+                self.ds = ds
+
+            a = phi * flux * dx
+            L = phi * rain * un * rho * self.ds
```
Comment (wence-) [inline gusto/diagnostics.py]: So this is diagnosing the flux into the top and out of the bottom?

File: gusto/physics.py [hunk 9]
```diff
@@ -104,6 +106,39 @@ def __init__(self, state):
...
+
+        # function spaces
+        Vt = self.rain.function_space()
+        Vu = state.fields('u').function_space()
+
+        # introduce sedimentation rate
+        # for now assume all rain falls at terminal velocity
+        terminal_velocity = 10  # in m/s
+        self.v = state.fields("rainfall_velocity", Vu)
+        self.v.project(as_vector([0, -terminal_velocity]))
```
Comment (wence-) [inline gusto/physics.py]: No need to make a whole function for this, just do:

```python
self.v = as_vector([0, -terminal_velocity])
```

Comment (wence-) [reply gusto/physics.py]: Oh, but then looking at the code, the `update_ubar` function requires fields, so OK.

File: gusto/transport_equation.py (hunk 7 above)
Comment (wence-) [reply gusto/transport_equation.py]: The advection term as coded doesn't add any surface integrals if `ibp` is `None`.  So I'm confused like Jemma.

Comment (wence-) [reply gusto/transport_equation.py]: FWIW, I note that `self.inflow` is not a thing that exists.  So is `outflow` really the right name?

Comment (wence-) [reply gusto/transport_equation.py]: I think this would somehow be clearer if the "outflow" term were somehow added separately.  But that would necessitate much larger redesign.

The idea would follow something like what thetis does:

You have `Term` objects that implement `a` and `L`.

Then `Equation` objects don't implement the terms directly, but rather you call `Equation.add_term(TermInstance)`.

So you would have:

```python
class Term(object):
   ...

class OutflowTerm(Term):
    def residual(...):
       pass

class AdvectionEquation(Equation):
    ...

eq = AdvectionEquation(...)
if outflow:
   eq.add_term(OutflowTerm(...))
```

Or something.

But this is food for thought, rather than to be done here.

Comment (tommbendall) [reply gusto/transport_equation.py]: The short answer is that this only works if `self.outflow=True` and `ibp is not None`. I'll just make this clearer by adding a warning if someone sets `ibp=None`.

What the `outflow` integral is trying to do is to add the contribution of any outgoing flux, but not incoming. This should be because `self.un` is `u` on exterior facets when `u` is pointing outwards, whilst `self.un=0` when `u` is pointing inwards.

You're right, there isn't such a thing as `self.inflow`, although we could add it in the same way if we wanted to.

With the integration by parts, I had thought that when `ibp=None`, the whole surface integral is there inside the main integral, `test*inner(u, grad(q))*dx`. When you integrate by parts, you separate out the surface integral to
`-q*div(test*u)*dx + test*q*inner(u,n)*dF`,
where `dF` is over all facets. We then separate the facets into being exterior and interior, and then normally we use boundary conditions to neglect the integral on the exterior facets. So I had presumed that if someone was choosing to use `ibp=None` then they would be trying to take include flow through the surfaces, and so wouldn't need to use the `outflow` option (which restores only outflow and not inflow).

Probably there is a mistake in something above that I said above and that's why it's confusing you!

File: gusto/diagnostics.py (hunk 8 above)
Comment (tommbendall) [reply gusto/diagnostics.py]: Yes, although I am expecting there only to be flux out and not in!

File: gusto/diagnostics.py [hunk 10]
```diff
@@ -401,6 +401,42 @@ def name(self):
...
+            flux = TrialFunction(space)
+            n = FacetNormal(state.mesh)
+            un = 0.5 * (dot(v, n) + abs(dot(v, n)))
+            self.flux = Function(space)
+            self.total_water = Function(space)
+
+            if space.extruded:
+                self.ds = ds_b + ds_t + ds_v
+            else:
+                self.ds = ds
```
[the remark this answers is not in the document]
Comment (tommbendall) [reply gusto/diagnostics.py]: I think this was done because otherwise you are redefining `firedrake.ds` which gives a lint error, and assigning it to be a member seemed a convenient enough way to do that.

[the remark this answers is not in the document]
Comment (wence-) [reply gusto/diagnostics.py]: OK, yes, I see (I hate python's scoping rules).

Yeah, either use a different variable name or else maybe write:

```python
            a = phi * flux * dx
            L = phi * rain * un * rho
            if space.extruded:
                L = L*(ds_b + ds_t + ds_v)
            else:
                L = L*ds
```

Or similar?

File: gusto/diagnostics.py [hunk 11]
```diff
@@ -438,6 +437,42 @@ def name(self):
...
+            a = phi * flux * dx
+            L = phi * rain * un * rho * self.ds
+
+            # setup solver
+            problem = LinearVariationalProblem(a, L, self.flux)
+            self.solver = LinearVariationalSolver(problem)
+
+    def compute(self, state):
+        self.solver.solve()
+        self.field.assign(self.total_water + self.flux)
```
Comment (wence-) [inline gusto/diagnostics.py]: I think this is wrong, `total_water` is always zero.

File: gusto/transport_equation.py [hunk 12]
```diff
@@ -181,6 +183,13 @@ def advection_term(self, q):
                             dot(self.ubar('+'), self.n('+'))*q('+'))
                       + inner(self.test('-'),
                               dot(self.ubar('-'), self.n('-'))*q('-')))*self.dS
+
+        if self.outflow:
+            if self.ibp is not None:
+                L += self.test*self.un*q*self.ds
+            else:
+                raise ValueError("ibp=None is not compatible with outflow option")
```
Comment (wence-) [inline gusto/transport_equation.py]: I think I would prefer the error checking to happen in `__init__`, then you can just write:
```python
if self.outflow:
    L += ...
```
here.

File: gusto/transport_equation.py [hunk 13]
```diff
@@ -149,16 +153,14 @@ class AdvectionEquation(TransportEquation):
...
         if equation_form == "advective" or equation_form == "continuity":
             self.continuity = (equation_form == "continuity")
         else:
             raise ValueError("equation_form must be either 'advective' or 'continuity'")
-        if vector_manifold:
-            self.vector_manifold = True
-        else:
-            self.vector_manifold = False
+        self.vector_manifold = vector_manifold
+        self.outflow = outflow
```
Comment (wence-) [inline gusto/transport_equation.py]: So check here with:
```python
if outflow and ibp is not None:
    raise ValueError(...)
```
